source/src/ble/cs_CrownstoneCentral.cpp

```diff
@@ -0,0 +1,663 @@
+	if (result.uuid == uuid) {
+		_sessionDataHandle = result.valueHandle;
+		LOGCsCentralDebug("Found session data handle: %u", _sessionDataHandle);
+	}
+	uuid.fromBaseUuid(_uuids[ServiceIndex::SERVICE_INDEX_CROWNSTONE], CONTROL_UUID);
+	if (result.uuid == uuid) {
+		_controlHandle = result.valueHandle;
+		LOGCsCentralDebug("Found control handle: %u", _controlHandle);
+	}
+	uuid.fromBaseUuid(_uuids[ServiceIndex::SERVICE_INDEX_CROWNSTONE], RESULT_UUID);
+	if (result.uuid == uuid) {
+		_resultHandle = result.valueHandle;
+		_resultCccdHandle = result.cccdHandle;
+		LOGCsCentralDebug("Found result handle: %u", _resultHandle);
+	}
+
+	// Setup mode
+	if (result.uuid == _uuids[ServiceIndex::SERVICE_INDEX_SETUP]) {
+		_opMode = OperationMode::OPERATION_MODE_SETUP;
+		LOGCsCentralInfo("Setup service found");
+	}
+	uuid.fromBaseUuid(_uuids[ServiceIndex::SERVICE_INDEX_SETUP], SETUP_KEY_UUID);
+	if (result.uuid == uuid) {
+		_sessionKeyHandle = result.valueHandle;
+		LOGCsCentralDebug("Found session key handle: %u", _sessionKeyHandle);
+	}
+	uuid.fromBaseUuid(_uuids[ServiceIndex::SERVICE_INDEX_SETUP], SESSION_DATA_UNENCRYPTED_UUID);
+	if (result.uuid == uuid) {
+		_sessionDataHandle = result.valueHandle;
+		LOGCsCentralDebug("Found session data handle: %u", _sessionDataHandle);
+	}
+	uuid.fromBaseUuid(_uuids[ServiceIndex::SERVICE_INDEX_SETUP], SETUP_CONTROL_UUID);
+	if (result.uuid == uuid) {
+		_controlHandle = result.valueHandle;
+		LOGCsCentralDebug("Found control handle: %u", _controlHandle);
+	}
+	uuid.fromBaseUuid(_uuids[ServiceIndex::SERVICE_INDEX_SETUP], SETUP_RESULT_UUID);
+	if (result.uuid == uuid) {
+		_resultHandle = result.valueHandle;
+		_resultCccdHandle = result.cccdHandle;
+		LOGCsCentralDebug("Found result handle: %u", _resultHandle);
+	}
+
+}
+
+void CrownstoneCentral::onDiscoveryDone(cs_ret_code_t retCode) {
+	if (!finalizeStep(ConnectSteps::DISCOVER, retCode)) {
+		return;
+	}
+	enableNotifications();
+}
+
+void CrownstoneCentral::onRead(ble_central_read_result_t& result) {
+	cs_ret_code_t retCode = result.retCode;
+	if (retCode != ERR_SUCCESS) {
+		finalizeOperation(_currentOperation, retCode);
+		return;
+	}
+
+	switch (_currentOperation) {
+		case Operation::CONNECT: {
+			ConnectSteps currentStep = static_cast<ConnectSteps>(_currentStep);
+			switch (currentStep) {
+				case ConnectSteps::SESSION_DATA: {
+					if (!finalizeStep(_currentStep, retCode)) {
+						return;
+					}
+					// Parse data that was read.
+					if (result.data.len != sizeof(session_data_t)) {
+						finalizeOperation(_currentOperation, ERR_WRONG_PAYLOAD_LENGTH);
+						return;
+					}
+					// Store the session data.
+					retCode = ConnectionEncryption::getInstance().setSessionData(*reinterpret_cast<session_data_t*>(result.data.data));
+					if (retCode != ERR_SUCCESS) {
+						finalizeOperation(_currentOperation, retCode);
+						return;
+					}
+					// Done.
+					finalizeOperation(_currentOperation, retCode);
+					break;
```

Comment:
maybe extract these into methods for readability :-)

Reply:
Moved all code in `case Operation::CONNECT` to a separate function, as I expect more Operation cases in the future (reading the device information service).